### PR TITLE
fix #297339, fix #46036, fix #297328: header clef issues [WIP]

### PR DESCRIPTION
**Pull request by Howard-C**
Resolves: https://musescore.org/node/297339.

Current code shows, if there's a clef at tick 0, then that clef's clef type becomes the default clef type for the staff.

This patch makes the code only do that if that clef is the header clef. This is judged by whether or not the clef segment is the first segment in the measure. If it is, then the variable `segment` still has the initial value 0; if not, then it has been assigned to a certain segment while dealing with the segments before.

Resolves: https://musescore.org/node/46036.

A clef is a header clef if and only if: it is at tick 0, and it is the first segment in measure.

### CODE REVIEW

File: libmscore/read206.cpp [hunk 1]
```diff
@@ -3001,7 +3001,7 @@ static void readMeasure(Measure* m, int staffIdx, XmlReader& e)
                   clef->setTrack(e.track());
                   clef->read(e);
                   clef->setGenerated(false);
-                  if (e.tick().isZero()) {
+                  if (e.tick().isZero() && clef->segment()) { // if segment(), then it's a header clef
```
Comment (dmitrio95) [inline libmscore/read206.cpp]: But `clef` hasn't been yet added anywhere, wouldn't this condition be always false?

Comment (Howard-C) [reply libmscore/read206.cpp]: Ah, it's true. But my test shows sometimes this works :O Don't know why.

Comment (Howard-C) [reply libmscore/read206.cpp]: So, um, what's the exact way to judge if a clef is header clef?

Comment (MarcSabatella) [reply libmscore/read206.cpp]: If it's already added to the score, then you check the segment type (CLEF vs HEADER_CLEF).  But if it's not added to the score, this is probably the wrong place to be making decisions on that basis.

Comment (Howard-C) [reply libmscore/read206.cpp]: Or is there a way to check if at the same tick, the clef is placed before the time signature? That can do the same thing.

Comment (Howard-C) [reply libmscore/read206.cpp]: Ah, we can simply check if the clef segment is the first one in the measure. See the new commit message.

File: mtest/libmscore/compat206/lidemptytext.mscx [hunk 2]
```diff
@@ -153,10 +153,6 @@
         </VBox>
       <Measure number="1">
         <stretch>0.9</stretch>
-        <Clef>
-          <concertClefType>G</concertClefType>
-          <transposingClefType>G</transposingClefType>
-          </Clef>
```
Comment (dmitrio95) [inline mtest/libmscore/compat206/lidemptytext.mscx]: Why did you remove this clef? It seems to be removed from the source file, not from the reference?

Comment (Howard-C) [reply mtest/libmscore/compat206/lidemptytext.mscx]: Hmmm, I'll look into that...

Comment (Jojo-Schmitz) [reply mtest/libmscore/compat206/lidemptytext.mscx]: Guess you'd need to add them to the ref file instead?

Comment (Howard-C) [reply mtest/libmscore/compat206/lidemptytext.mscx]: No it's not that simple. There isn't supposed to be changes in these two files, so I guess I made some pretty harmful deletions back then.

**Comment by vpereverzev** [comment]
@Howard-C Looks like PR is hanging
Is it still actual for master/3.x?

**Comment by Howard-C** [comment]
No, the fix is incorrect. Maybe I'll revisit this later

**Comment by AntonioBL** [comment]
@Howard-C :
Ciao.
Do you think all the fixes you implemented are incorrect? If you let me, I can try to look at your fix and port them to current 3.x branch.

**Comment by Howard-C** [comment]
Not all are incorrect, it is one of the test cases that it cannot correctly process. See https://github.com/musescore/MuseScore/pull/5484/files#diff-e730ce47006c25eb38b2f40a553b48409c913b27498e3c3da12142b4bbe86f76, I have to change the source file, not the reference, so it's not correct

**Comment by AntonioBL** [comment]
Ciao.
I looked at the proposed solution, and I found it correctly working in every case I tested.
Only the mtest should be updated for the new parameters, and the ref instead of the basic file should be modified for lidemptytext.mscx: that clef is not necessary (it is the same as the default clef for that staff), but it should be correct syntax, if I understood correctly, and MuseScore displays the same result for both cases with or without that clef.

**Comment by Howard-C** [comment]
So this kind of "optimisation" can never be wrong? If you believe so, feel free to create a PR :) I'm very busy recently.
